### ISSUE TITLE
Add support for strict mode during tuist project generation

### ISSUE THREAD

**Issue by nicholaskim94** [tuist/tuist#4376]
## Context 🕵️‍♀️
Currently, lint warnings generated during `tuist generate` command does not result in command failure. 
However, because logs generated by the generate command tends to follow the whole workspace size, it's easy to miss the linting warnings generated by the linters, and therefore lead to potential errors. 
Currently experiencing some runtime issues due to `StaticProductsGraphLint`s not failing during project generation.

## Proposal 🎉
Add a `--strict` flag to `tuist generate` command to make the command fail entirely when any linting issues of either type .warning or .error are generated

Any feedback would be greatly appreciated! Thanks.

**Comment by kwridan** [comment]
Thanks for proposing this @nicholaskim94 

I think this would be a great addition 👍 

An additional suggestion I would add is deferring the display of warnings till after generation - this may also help surface warning even in the non strict case for larger workspaces.

```sh
tust generate
Resolved cache profile 'Development' from Tuist's defaults
Generating workspace Workspace.xcworkspace
Generating project MainApp
Generating project Framework1
Generating project Framework2

Warning: ....

Project generated.
Total time taken: 0.278s
```

_The small caveat here is, the lint process does need to take place ahead of generation, but we can defer the results display till after generation_

Would you like to contribute either of those changes?  Happy to guide you through this :)

**Comment by nicholaskim94** [comment]
@kwridan Thanks for the feedback! 
I'll try to come up with a PR including both changes

**Comment by fortmarek** [comment]
If you want to enforce stricter rules, it might be better to share that configuration across your team, so I wonder if we should put this inside `Config.swift` instead of the flag.

**Comment by danyf90** [comment]
Since those are warnings, if we add a strict mode we should probably add some way to exclude some lint warnings to make the generate fail, otherwise as soon as you have one case that generates a warning but you want to keep it for reasons, then you can't use strict anymore

**Comment by fortmarek** [comment]
> otherwise as soon as you have one case that generates a warning but you want to keep it for reasons, then you can't use strict anymore

I'd argue that if the lint rule is not valid, we should change it, or the team should fallback to using the non-strict version. Opting out of individual lint rules is imho unnecessarily complicated and I would rather keep this simple.

**Comment by danyf90** [comment]
> I'd argue that if the lint rule is not valid, we should change it

If it's a warning, I would say that in some scenarios it could make sense to be ignored, otherwise it should be an error also without the need to add the `--strict` flag

> otherwise as soon as you have one case that generates a warning but you want to keep it for reasons, then you can't use strict anymore

I agree, my point was that just adding `--strict` could not be enough in some cases, so I am leaning more towards Kas proposal of making the warnings more visible

**Comment by fortmarek** [comment]
> I am leaning more towards Kas proposal of making the warnings more visible

I think this should be the first step - but still, having something akin to the `--strict` flag would still be useful to ensure that no warnings are emitted on CI (and make the pipeline fail if there are any)

**Comment by ezraberch** [comment]
> If you want to enforce stricter rules, it might be better to share that configuration across your team, so I wonder if we should put this inside `Config.swift` instead of the flag.


> Since those are warnings, if we add a strict mode we should probably add some way to exclude some lint warnings to make the generate fail, otherwise as soon as you have one case that generates a warning but you want to keep it for reasons, then you can't use strict anymore

If we do it in Config.swift, we could allow passing of a closure which takes a warning and returns whether the warning should be ignored, be printed, or abort.

**Comment by fortmarek** [comment]
> If we do it in Config.swift, we could allow passing of a closure which takes a warning and returns whether the warning should be ignored, be printed, or abort.

Yeah, it's something we can consider in the next iteration 👀

**Comment by danyf90** [comment]
> > If we do it in Config.swift, we could allow passing of a closure which takes a warning and returns whether the warning should be ignored, be printed, or abort.
> 
> Yeah, it's something we can consider in the next iteration 👀

Doesn't Config need to be "static" (i.e. how would you generate the JSON to parse it)?

**Comment by fortmarek** [comment]
> Doesn't Config need to be "static" (i.e. how would you generate the JSON to parse it)?

That is true, `Config` needs to conform to `Codable`. However, such a feature can be implemented without using callbacks but rather using simple structs or enums.

**Comment by nicholaskim94** [comment]
I've come up with a PR containing @kwridan 's idea of deferring warnings. 
I might need some guidance as it's my first code PR on this project :)

**Comment by danyf90** [comment]
Closing as it has been "solved" by printing the warnings at the end of the generation

**Comment by fortmarek** [comment]
Reopening as we are reconsidering this and even with warnings at the end of the generation, the `--strict` flag may be useful, especially on the CI.

**Comment by romeouald** [comment]
Hey @fortmarek & everyone! I was thinking about picking up adding the `--strict` flag. What would be the preference here?
- Only add it to the `generate` command
- Add a global flag, because under the hood multiple commands could initiate the generation (eg.: `BuildService`, `RunService`, `TestService`)

Any feedback is welcome, thanks in advance!

**Comment by pepicrft** [comment]
That'd be amazing @romeouald 🚀. I'm inclined to suggest that we add it to all the commands that use generation under the hood.  Thoughts @tuist/community-core ?

**Comment by fortmarek** [comment]
> I'm inclined to suggest that we add it to all the commands that use generation under the hood. Thoughts @tuist/community-core ?

Yes, that makes sense to me 👍
